src/ImageSharp/Formats/Png/Filters/AverageFilter.cs

```diff
@@ -79,6 +84,89 @@ public static void Encode(Span<byte> scanline, Span<byte> previousScanline, Span
                 sum += Numerics.Abs(unchecked((sbyte)res));
             }
 
+#if SUPPORTS_RUNTIME_INTRINSICS
+            if (Avx2.IsSupported)
+            {
+                Vector256<int> sumAccumulator = Vector256<int>.Zero;
+
+                for (int xLeft = x - bytesPerPixel; x + Vector256<byte>.Count <= scanline.Length; xLeft += Vector256<byte>.Count)
+                {
+                    Vector256<byte> scan = Unsafe.As<byte, Vector256<byte>>(ref Unsafe.Add(ref scanBaseRef, x));
+                    Vector256<byte> left = Unsafe.As<byte, Vector256<byte>>(ref Unsafe.Add(ref scanBaseRef, xLeft));
+                    Vector256<byte> above = Unsafe.As<byte, Vector256<byte>>(ref Unsafe.Add(ref prevBaseRef, x));
+
+                    Vector256<byte> res = Avx2.Subtract(scan, Average(left, above));
+                    Unsafe.As<byte, Vector256<byte>>(ref Unsafe.Add(ref resultBaseRef, x + 1)) = res; // +1 to skip filter type
+                    x += Vector256<byte>.Count;
+
+                    Vector256<sbyte> absRes = Avx2.Abs(res.AsSByte()).AsSByte();
+                    Vector256<short> loRes16 = Avx2.UnpackLow(absRes, Vector256<sbyte>.Zero).AsInt16();
+                    Vector256<short> hiRes16 = Avx2.UnpackHigh(absRes, Vector256<sbyte>.Zero).AsInt16();
+
+                    Vector256<int> loRes32 = Avx2.UnpackLow(loRes16, Vector256<short>.Zero).AsInt32();
+                    Vector256<int> hiRes32 = Avx2.UnpackHigh(loRes16, Vector256<short>.Zero).AsInt32();
+                    sumAccumulator = Avx2.Add(sumAccumulator, loRes32);
+                    sumAccumulator = Avx2.Add(sumAccumulator, hiRes32);
+
+                    loRes32 = Avx2.UnpackLow(hiRes16, Vector256<short>.Zero).AsInt32();
+                    hiRes32 = Avx2.UnpackHigh(hiRes16, Vector256<short>.Zero).AsInt32();
+                    sumAccumulator = Avx2.Add(sumAccumulator, loRes32);
+                    sumAccumulator = Avx2.Add(sumAccumulator, hiRes32);
+                }
+
+                for (int i = 0; i < Vector256<int>.Count; i++)
+                {
+                    sum += sumAccumulator.GetElement(i);
+                }
```

Comment:
This can be [vectorized](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAJI8ovLrl5hcAbhr0mrTj36CRGMRKm4WADQAcSOdXkBmJqQYBhBgG8aDR03PiMDAIIAKAGowwGaFIAViQAHlcAPgZcDnx3MDBYjgAbbACoAEoGByd7aicChldo2IYAXgY6Y0KnHJqAM2gGT2LecsqZIoZQhl9/QJDwngiWKwhuDE6AaineDLrCvJqamPwGKYrV+MT8FLToFgBxGAwAUWSYAR4WjOrlgF8FgqenYgB2Evw7x0f82r/HMQXDwGAAhHx+dLBMKRT7bJKpdLzAF2F6OPpQwawgAW2AAJnj2u4AG4IUgsAAS0F4AC8IDxsMl3ATPFsEgj9lA0HD2btEdBbmjepDoORSPohhgoskIO1cQSjicADIQADuMCgnkFKIKMvaAGVcLgYKZKdS6QymSyZdyZdrlsL+lAxRKcW0KvK8YqMABVAAOfo1Wu+hWx7oYhuNpqpYgtGEZzLxnjD3LD9uWGNF4slUQw+D9BuN5MTnhtDDTIYK7wYeb9LAAKhB9WBGdhNemfjR7kA==) too.

```suggestion
                Vector256<int> hadd = Avx2.HorizontalAdd(sumAccumulator, sumAccumulator);
                Vector128<int> lo = hadd.GetLower();
                lo = Ssse3.HorizontalAdd(lo, lo);
                Vector128<int> hi = hadd.GetUpper();
                hi = Ssse3.HorizontalAdd(hi, hi);
                Vector128<int> tmp = Sse2.Add(lo, hi);
                sum += tmp.ToScalar();
```

Reply:
Sounds like a good idea but this barely had any effect on performance as it's only called once per row.

Reply:
> it's only called once per row

Yep, but it's less produced machine-code which is almot always better in real-workloads (even it doesn't show up in a _micro_-benchmark.

Reply:
`phaddd` is implemented in microcode on all supporting processors, so it's slow compared to `pshufd`+`paddd`, unless of course you're actually adding horizontal pairs in two vectors *and* packing the result into one.

This can be written in smaller code as:

```C#
Vector128<int> vsum = Sse2.Add(sumAccumulator.GetLower(), sumAccumulator.GetUpper()); // add upper lane to lower lane
vsum = Sse2.Add(vsum, Sse2.Shuffle(vsum, 0b_11_11_01_01));                            // add odd to even
vsum = Sse2.Add(vsum, Sse2.Shuffle(vsum, 0b_11_10_11_10));                            // add high to low
sum += Sse2.ConvertToInt32(vsum);                                                     // Vector128<int>.ToScalar() isn't optimized pre-net5.0 https://github.com/dotnet/runtime/pull/37882
```

Reply:
Good to know -- thanks!